### ISSUE TITLE
sql: inserting with default values for both INT ARRAY and STRING ARRAY columns fails

### ISSUE THREAD

**Issue by codise** [cockroachdb/cockroach#38293]
**Describe the problem**

When adding rows using default values to tables with `INT[] NOT NULL` and `STRING[] NOT NULL` columns (default=empty array) the insert fails.

The error message is either:

- `pq: type of array contents int doesn't match column type STRING`
- or `pq: type of array contents string doesn't match column type INT`.

depending which column is defined first. So the second column will produce the error for some reason.

If there are two array columns of same type this issue doesn't seem to occur.

**To Reproduce**

Case 1:
```
CREATE TABLE IF NOT EXISTS bugtest1 (
    id SERIAL NOT NULL PRIMARY KEY,
    arr1 STRING(100) ARRAY NOT NULL default ARRAY[],
    arr2 INT ARRAY NOT NULL default ARRAY[]
);
insert into bugtest1 (id) values (1);
```
Result:
`pq: type of array contents string doesn't match column type INT`

Case 2:
```
CREATE TABLE IF NOT EXISTS bugtest2 (
    id SERIAL NOT NULL PRIMARY KEY,
    arr2 INT ARRAY NOT NULL default ARRAY[],
    arr1 STRING(100) ARRAY NOT NULL default ARRAY[]
);
insert into bugtest2 (id) values (1);
```
Result:
`pq: type of array contents int doesn't match column type STRING`

**Expected behavior**

DB should use the default value provided and insert to row successfully. In previous version I used (2.x.x) this or at least comparable example used to work.

**Additional data / screenshots**

**Environment:**

 - CockroachDB version: CockroachDB CCL v19.1.2 (x86_64-apple-darwin18.6.0, built 2019/06/17 20:21:54, go1.12. (installed using brew)
 - Server OS: Mac OS X 10.14.4
 - Client app: `cockroach sql`

**Additional context**

Running a migration tools to get the database up to date now fails because earlier migration would assume this syntax works (and it worked for a long time). Default values for array types seem to be very broken in this version.

This issue could be somewhat related: [Issue 23429](https://github.com/cockroachdb/cockroach/issues/23429)

**Comment by jordanlewis** [comment]
Hi @codise,

Thanks very much for the report and excellent reproduction steps.

Here's some more detail for the optimizer team:

```
root@127.0.0.1:55198/defaultdb> CREATE TABLE IF NOT EXISTS bugtest1 (                                                                                                                                           id SERIAL NOT NULL PRIMARY KEY,                                                                                                                                                                             arr1 STRING(100) ARRAY NOT NULL default ARRAY[],                                                                                                                                                            arr2 INT ARRAY NOT NULL default ARRAY[]                                                                                                                                                                 );                                                                                                                                                                                                          insert into bugtest1 (id) values (1);
CREATE TABLE

Time: 2.364ms

pq: type of array contents string doesn't match column type IntFamily
root@127.0.0.1:55198/defaultdb> set optimizer=off;                                                                                                                                                          SET

Time: 225µs

root@127.0.0.1:55198/defaultdb> insert into bugtest1 (id) values (1);
INSERT 1

Time: 750µs

root@127.0.0.1:55198/defaultdb> explain insert into bugtest1 (id) values (1);
        tree       |  field   |       description
+------------------+----------+--------------------------+
  count            |          |
   └── insert      |          |
        │          | into     | bugtest1(id, arr1, arr2)
        │          | strategy | inserter
        └── values |          |
                   | size     | 3 columns, 1 row
(6 rows)

Time: 454µs

root@127.0.0.1:55198/defaultdb> set optimizer=on;
SET

Time: 276µs

root@127.0.0.1:55198/defaultdb> explain insert into bugtest1 (id) values (1);
          tree          |  field   |       description
+-----------------------+----------+--------------------------+
  count                 |          |
   └── insert           |          |
        │               | into     | bugtest1(id, arr1, arr2)
        │               | strategy | inserter
        └── render      |          |
             └── values |          |
                        | size     | 2 columns, 1 row
(7 rows)

Time: 536µs

root@127.0.0.1:55198/defaultdb>
```

It looks to me like the optimizer is doing something slightly wrong in the presence of multiple defaults. @RaduBerinde could you please find someone to take a look at this, or else kick it back if you can determine that the optimizer's doing the right thing here?

**Comment by RaduBerinde** [comment]
Sure looks like the optimizer.

**Comment by RaduBerinde** [comment]
The optimizer generates strings to deduplicate expressions. It uses `tree.AsStringWithFlags(expr, tree.FmtCheckEquivalence)`. This just returns `ARRAY[]` regardless of the resolved type.
```
expr: &tree.Array{Exprs:tree.Exprs(nil), typeAnnotation:tree.typeAnnotation{typ:(*types.T)(0xc0009419e0)}}  str: ARRAY[]
expr: &tree.Array{Exprs:tree.Exprs(nil), typeAnnotation:tree.typeAnnotation{typ:(*types.T)(0xc000941a70)}}  str: ARRAY[]
```

**Comment by RaduBerinde** [comment]
@codise I have a fix out, it should make it into the next 19.1 release. In the meantime, you could work around by adding a type cast to the `ARRAY[]`:

```
CREATE TABLE t (
    id SERIAL NOT NULL PRIMARY KEY,
    STRING(100) ARRAY NOT NULL default ARRAY[]::STRING(100)[],                                                                                                                                         
    arr2 INT ARRAY NOT NULL default ARRAY[]::INT[]
);
```

**Comment by codise** [comment]
Thank you for super fast and professional bug fixing and a workaround.
